### PR TITLE
Initial unit tests for SVT-AV1 encoder

### PR DESCRIPTION
**Pull request by Cidana-Developers**
1. Added end-to-end test, which will execute conformance test
2. Added unit tests for modules including tx, quant/dequant,
entropy coding;
3. Added api test, which will check with different parameters

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -28,6 +37,11 @@ if (MSVC OR MSYS OR MINGW OR WIN32)
 option(BUILD_SHARED_LIBS "Build shared libraries (DLLs)." ON)
 endif()
 
+if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
```
Comment (1480c1) [inline test/CMakeLists.txt]:
```suggestion
if(MSVC)
```
Please use msvc instead of matching windows since this will also affect mingw and other compilers that does not accept `/*` style flags

File: test/api_test/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+# Include Subdirectories
+include_directories (${PROJECT_SOURCE_DIR}/test/)
+include_directories (${PROJECT_SOURCE_DIR}/Bin/${CMAKE_BUILD_TYPE}/)
+include_directories(${PROJECT_SOURCE_DIR}/third_party/googletest/include third_party/googletest/src)
+include_directories(${PROJECT_SOURCE_DIR}/Source/API)
+
+# Define helper functions and macros used by Google Test.
+include(../../third_party/googletest/cmake/internal_utils.cmake)
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
```
Comment (1480c1) [inline test/api_test/CMakeLists.txt]:
```suggestion
if(MSVC)
```
Ditto here.

File: test/e2e_test/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+if (APPLE)
+link_directories(${PROJECT_SOURCE_DIR}/third_party/aom/lib/mac)
+endif(APPLE)
+
+
+# Define helper functions and macros used by Google Test.
+include(../../third_party/googletest/cmake/internal_utils.cmake)
+
+if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
```
Comment (1480c1) [inline test/e2e_test/CMakeLists.txt]:
```suggestion
if(MSVC)
```

**Comment by 1480c1** [comment]
Can you also add a separate job for the unit tests? It seems that a lot of them failed but since it's in the coveralls section, it does not mark it as failed.

**Comment by 1480c1** [comment]
Something like:
``` diff
diff --git a/.travis.yml b/.travis.yml
index c6cae3d..b88951d 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -114,7 +114,7 @@ matrix:
       stage: Coveralls
       os: linux
       compiler: gcc
-      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON"
+      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON -DBUILD_TESTING=ON"
       script:
         - *base_script
         - *coveralls_script
@@ -123,7 +123,7 @@ matrix:
       stage: Coveralls
       os: osx
       compiler: clang
-      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON"
+      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON -DBUILD_TESTING=ON"
       script:
         - *base_script
         - *coveralls_script
@@ -136,3 +136,15 @@ matrix:
       script:
         - *base_script
         - valgrind -- SvtAv1EncApp -enc-mode 8 -w 720 -h 486 -fps 60 -i akiyo_cif.y4m -n 150 -b test1.ivf
+    - name: Unit Tests
+      stage: test
+      compiler: gcc
+      os: linux
+      env: build_type=debug CMAKE_EFLAGS="-DBUILD_TESTING=ON"
+      script:
+        - *base_script
+        - cd $TRAVIS_BUILD_DIR/Build/linux/$build_type
+        - cmake -j $(if [ $TRAVIS_OS_NAME = osx ]; then sysctl -n hw.ncpu; else nproc; fi) --build . --target SvtAv1UnitTests;SvtAv1ApiTests;SvtAv1E2ETests;TestVectors
+        - SvtAv1UnitTests
+        - SvtAv1ApiTests
+        - SvtAviE2ETests
\ No newline at end of file

```

also,

**Comment by kodabb** [comment]
ah another giant PR, another example for #238

**Comment by RyanZhao-Cidana** [comment]
> Something like:
> 
> ```diff
> diff --git a/.travis.yml b/.travis.yml
> index c6cae3d..b88951d 100644
> --- a/.travis.yml
> +++ b/.travis.yml
> @@ -114,7 +114,7 @@ matrix:
>        stage: Coveralls
>        os: linux
>        compiler: gcc
> -      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON"
> +      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON -DBUILD_TESTING=ON"
>        script:
>          - *base_script
>          - *coveralls_script
> @@ -123,7 +123,7 @@ matrix:
>        stage: Coveralls
>        os: osx
>        compiler: clang
> -      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON"
> +      env: COVERALLS_PARALLEL=true build_type=debug CMAKE_EFLAGS="-DCOVERAGE=ON -DBUILD_TESTING=ON"
>        script:
>          - *base_script
>          - *coveralls_script
> @@ -136,3 +136,15 @@ matrix:
>        script:
>          - *base_script
>          - valgrind -- SvtAv1EncApp -enc-mode 8 -w 720 -h 486 -fps 60 -i akiyo_cif.y4m -n 150 -b test1.ivf
> +    - name: Unit Tests
> +      stage: test
> +      compiler: gcc
> +      os: linux
> +      env: build_type=debug CMAKE_EFLAGS="-DBUILD_TESTING=ON"
> +      script:
> +        - *base_script
> +        - cd $TRAVIS_BUILD_DIR/Build/linux/$build_type
> +        - cmake -j $(if [ $TRAVIS_OS_NAME = osx ]; then sysctl -n hw.ncpu; else nproc; fi) --build . --target SvtAv1UnitTests;SvtAv1ApiTests;SvtAv1E2ETests;TestVectors
> +        - SvtAv1UnitTests
> +        - SvtAv1ApiTests
> +        - SvtAviE2ETests
> \ No newline at end of file
> ```
> 
> also, add build_testing to the cmake flags

Work on this issue whole day, sad.
Here an issue, some test cased do failed with current SVT-AV1 encoder, 
and than gtest-based program will return the count of failed test item as return value, 
than travis-ci get the return value which ">0", 
than travis-ci think this build failed.

Following the guide https://docs.travis-ci.com/user/job-lifecycle#complex-build-commands
add a bash script to run test program, and use 'exit 0' in bash script to output a zero return value.

File: test/CMakeLists.txt (hunk 1 above)
Comment (Cidana-Developers) [reply test/CMakeLists.txt]: Done.

File: test/api_test/CMakeLists.txt (hunk 2 above)
Comment (Cidana-Developers) [reply test/api_test/CMakeLists.txt]: Done.

File: test/e2e_test/CMakeLists.txt (hunk 3 above)
Comment (Cidana-Developers) [reply test/e2e_test/CMakeLists.txt]: Done.

**Comment by Cidana-Developers** [comment]
> Can you also add a separate job for the unit tests? It seems that a lot of them failed but since it's in the coveralls section, it does not mark it as failed.
> https://travis-ci.com/Cidana-Developers/SVT-AV1/jobs/201454816#L7209

Sure, will do it.

**Comment by kylophone** [comment]
> ah another giant PR, another example for #238

Yeah, big initial import, but it's pretty self contained. @Cidana-Developers, I think these will be more incremental going forward. Is that right?

**Comment by Cidana-Developers** [comment]
> > ah another giant PR, another example for #238
> 
> Yeah, big initial import, but it's pretty self contained. @Cidana-Developers, I think these will be more incremental going forward. Is that right?

Yes, no problem.
